server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,590 @@
+                                    Hierarchy.BAD_PREFIXES, controller, path
+                                )
+                            elif prefix_file.exists():
+                                rhier.add_unexpected_controllers(
+                                    Hierarchy.UNUSED_PREFIX_FILES, controller, path
+                                )
+                        else:
+                            if prefix:
+                                if prefix != prefix_path:
+                                    rhier.add_unexpected_controllers(
+                                        Hierarchy.BAD_PREFIXES, controller, path
+                                    )
+                            elif not prefix_file.exists():
+                                rhier.add_unexpected_controllers(
+                                    Hierarchy.MISSING_PREFIX_FILES, controller, path
+                                )
+                            else:
+                                f = 0
+                                try:
+                                    with prefix_file.open(mode="r") as file:
+                                        prefix = file.read().replace("\n", "")
+                                except Exception:
+                                    f = 1
+                                if f == 1:
+                                    rhier.add_unexpected_controllers(
+                                        Hierarchy.BAD_PREFIX_FILES, controller, path
+                                    )
+                                else:
+                                    if prefix != prefix_path:
+                                        rhier.add_unexpected_controllers(
+                                            Hierarchy.BAD_PREFIXES, controller, path
+                                        )
+                        verify_user_arg(rhier, mdlogcfg, controller, path)
+    return 0
+
+
+# Controller Hierarchy
+def verify_controllers(hier):
+    """ 'hier' is a ControllerHierarchy class object. """
+    controllers = Path(hier.path).glob("*")
+    verify_valid_controllers(hier, controllers)
+
+    if hier.controllers:
+        for controller in hier.controllers:
+            dir_path = Path(hier.path, controller)
+            unexpected_dirs = list()
+            if not Path(hier.config.ARCHIVE, controller).is_dir():
+                """We have a controller in the hierarchy which does not have a
+                controller of the same name in the archive hierarchy.  All
+                we do is report it, don't bother analyzing it further.
+                """
+                hier.add_controller_list(Hierarchy.MIALIST, Path(controller).name)
+            else:
+                """Report any controllers with objects other than directories
+                and links, while also recording any empty controllers.
+                """
+                if len(os.listdir(dir_path)) == 0:
```

Comment:
A couple of problems here:

- you do a `listdir()` here and then a `glob("*")` of the same directory on l. 413: since you need the results of the glob, do that first and then check if the resulting list is empty - if it is you mark it as an empty controller; if not, you go through the `for` loop (or as suggested in the meeting, you can do it in the other order: the nice thing about `for` loops is that they do nothing if the iterable is empty). 

- there is a difference between incoming and results hierarchy: a controller in the incoming hierarchy can have subdirectories but no symlinks; a controller in the results hierarchy can have subdirectories (because the user specified a prefix for the result) or symlinks (to the unpacked tar ball directories in the incoming hierarchy). So the test on l.415 is wrong: it should check different things in the two cases.

- The handling of `unexpected_dirs` is ... unexpected: there is a name problem in that, depending on the case, the unexpected objects can be a bunch of diferent things (links, special files etc) - in fact, whatever they are, they are not directories: directories are acceptable in either case.  In addition, AFAICT, we don't do anything with the value of `unexpected_dirs`: we collect them in the loop, we check whether the value is non-empty and then we throw it away. If we don't care about the exact value, we might as well get rid of it: in the loop, we do the `UNEXPECTED_CONTROLLERS` thing and break immediately: as soon as we find one bad entry, we don't need to go through all the rest of the entries (although again, there is a name problem: it should probably be `UNEXPECTED_OBJECTS_IN_CONTROLLER`).

Reply:
In general, we need to consider `os.listdir()` calls and `glob("*")` calls as being expensive and work to make sure we perform them only once where possible.